x-pack/plugins/lens/public/trigger_actions/view_underlying_data_action.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+import type { IEmbeddable } from 'src/plugins/embeddable/public';
+import { createAction } from '../../../../../src/plugins/ui_actions/public';
+import { Embeddable } from '../embeddable';
+import type { DiscoverStart } from '../../../../../src/plugins/discover/public';
+
+const ACTION_VIEW_UNDERLYING_DATA = 'ACTION_VIEW_UNDERLYING_DATA';
+
+export const viewUnderlyingDataAction = (discover: DiscoverStart) =>
+  createAction<{ embeddable: IEmbeddable }>({
+    type: ACTION_VIEW_UNDERLYING_DATA,
+    id: ACTION_VIEW_UNDERLYING_DATA,
+    getDisplayName: () => 'Open in Discover',
+    isCompatible: async (context: { embeddable: IEmbeddable }) => {
+      let isCompatible = false;
+      if (context.embeddable instanceof Embeddable) {
+        isCompatible = await context.embeddable.getCanViewUnderlyingData();
```

Comment:
I agree that we shouldn't calculate this on every active data change like Marco said. I think we can go with this:
* If `isCompatible` is called (which will happen if the user tries to open the context menu), check whether active data is available on the embeddable
* If not, return `false`
* If yes, get calculate the underlying data args to check whether they _can_ be compatible
* On execute use the cached underlying data args from the embeddable

This means the context menu won't wait for the current request to complete (which can take quite a while and maybe the user just wants to change the title which should be possible even during loading)